### DISCUSSION TITLE
Reconstruct a weighted adjacent matrix

### DISCUSSION THREAD

**Discussion by yjgeno** [pyg-team/pytorch_geometric#4513] [Q&A]
VGAE by default will reconstruct an adjacent matrix and the entries of which are simply the inner products of node latent embeddings.

I was wondering if there are already some efforts that allow the model to input and reconstruct a weighted adjacent matrix (which may contain pos/neg values)?
Any resources are appreciated.

**Comment by rusty1s** [comment]
How about you use an `MLP` on top of concatenated source and destination node embeddings to learn the weight between these nodes?
```python
x = GNN(x, edge_index)
x_src, x_dst = x[edge_index[0]], x[edge_index[1]]
edge_weight = MLP(torch.cat([x_src, x_dst], dim=-1))
```
For undirected (but weighted graphs), you may want to apply the Hadamard product (`x_src * x_dst`) instead of concatenation.

**Reply by yjgeno** [reply]
Hi @rusty1s thanks

If I understand correctly, you are suggesting replacing the default inner-product decoder of VGAE with an MLP (concat or get Hadamard product of embeddings of a pair of nodes and reconstruct their edge)?

**Reply by rusty1s** [reply]
Yes, indeed.

**Reply by yjgeno** [reply]
@rusty1s Thanks a lot
and looks like no papers discussing this
Please let me know if you know any resources already achieved weighted graph reconstruction

**Reply by yikang7720** [reply]
Hi @rusty1s, it is an extraordinary idea. Could you please tell me what is the source of the newly VGAE architecture? Like which paper indicated this idea?

**Reply by rusty1s** [reply]
I don't have a specific paper in mind for this. The closest to my mind would be the Node2Vec paper as they are discussing different ways to combine node-level representations.
